i18n/i18n.go

```diff
@@ -0,0 +1,95 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package i18n provides an implementation agnostic API to be used in
+// both library packages and application packages importing this library.
+// The application has the options to:
+//
+// 1. Initialise this package by implementing the i18n interface
+// 2. Leave the package uninitialised, disabling translation
+//
+// Warning:
+//
+// None of the i18n functionality may be used during early
+// initialisation code such as when defining a package 'const', 'var'
+// or 'init()'. This will prevent the application initialising
+// the i18n interface before it gets used in packages, and will
+// result in translation being disabled until the initialisation
+// call is made.
+
+package i18n
+
+import "github.com/Xuanwo/go-locale"
+
+// i18nIfc defines the interface that provides internationalisation
+// translations for packages. An implementation specific instance
+// of interface must be provided by the application wanting to use
+// translation.
+type i18nIfc interface {
+	CurrentLocale() string
+	G(msgid string) string
+	NG(msgid string, msgidPlural string, n int) string
+}
+
+var i18nInstance i18nIfc
+
+// Initialise must be called before any of the calls to i18n
+// functionality can be used by any packages inside both the library
+// or the application using the library.
+func Initialise(instance i18nIfc) {
+	i18nInstance = instance
```

Comment:
I'm a bit uncomfortable with this global and the global `i18n.Initialise` function, but maybe I'm missing some context. Are we always going to use gettext in a global way in utilities that use `golib`? It seems like `go-gettext` itself doesn't suffer from this "global" problem (you create a `TextDomain`, then get a locale/`Catalog`, then call `Gettext` et al on that). But I guess `snapd` does use the global approach in its `i18n` package.

So maybe global is fine for this use case. But it still seems awkward to require `i18n.Initialize` to be called in a certain order (which as you show, prevents the use of `i18n.G` at init time). Is there a better way to handle this?

Happy to get on a video call if it's easier to bring me up to speed / hash this out.

Reply:
Would love to see if you have a better solution to the problem.